README.md
[![Build Status](https://travis-ci.org/bitjson/geyser.svg?branch=master)](https://travis-ci.org/bitjson/geyser)
[![Codecov](https://img.shields.io/codecov/c/github/bitjson/geyser.svg)](https://codecov.io/gh/bitjson/geyser)
[![NPM version](https://img.shields.io/npm/v/geyser-mongo.svg)](https://www.npmjs.com/package/geyser-mongo)
[![Standard Version](https://img.shields.io/badge/release-standard%20version-brightgreen.svg)](https://github.com/conventional-changelog/standard-version)
[![dependencies Status](https://david-dm.org/bitjson/geyser/status.svg)](https://david-dm.org/bitjson/geyser)
[![devDependencies Status](https://david-dm.org/bitjson/geyser/dev-status.svg)](https://david-dm.org/bitjson/geyser?type=dev)

# Geyser

Publish and subscribe to streams of data over MongoDB.

## Example

In the source Node.js process, create a MongoObserver which writes to a `ticks` collection in MongoDB. Then subscribe it to a data source (in this example, an interval).

```js
import { Observable } from 'rxjs';
import { MongoObserver } from 'geyser-mongo';

const ticks = MongoObserver.create({
  db: 'mongodb://localhost/my-db',
  collection: 'ticks'
});

Observable.interval(1000).take(5).subscribe(ticks);
```

In the listening Node.js process, create a MongoObservable which emits messages written to the "ticks" collection.

```js
import { MongoObservable } from 'geyser-mongo';

const ticks = MongoObservable.create({
  db: 'mongodb://localhost/my-db',
  collection: 'ticks'
}).subscribe({
  next: console.log,
  complete: () => console.log('Complete.'),
  error: console.error
});

// 0
// 1
// 2
// 3
// 4
// Complete.
```

Each time a value is delivered to the MongoObserver, the MongoSubscriber will emit. The observable will also emit `error` and `complete` signals.

### Using an Existing MongoDB Connection

`MongoObservable` and `MongoObserver` can also be created from an existing MongoDB connection. The following example is a simple latency test.

```js
import { MongoClient } from "mongodb";
import { Observable } from 'rxjs';
import { MongoObservable, MongoObserver } from 'geyser-mongo';

MongoClient.connect(url).then(db => {
  const ticksIn = MongoObserver.create({
    collection: "ticks",
    db
  });

  const ticksOut = MongoObservable.create({
    collection: "ticks",
    db
  });

  Observable.interval(1000)
    .take(60)
    .map(i => i.toString())
    .startWith("connect")
    .do(i => console.time(i))
    .subscribe(ticksIn);

  ticksOut.subscribe({
    complete: () => db.close(),
    next: (i: string) => console.timeEnd(i)
  });
}
```

This example also demonstrates MongoObserver's internal buffering before it's connected to the database.

### Connecting the MongoObserver Early

Internally, MongoObserver calls it's `connect` method when it receives it's first event. It's possible to `connect` the MongoObserver early to reduce internal buffering.

```js
import { Observable } from 'rxjs';
import { MongoObserver } from 'geyser-mongo';

const ticks = MongoObserver.create({
  db: 'mongodb://localhost/my-db',
  collection: 'ticks'
}).connect(console.error, () => console.log("MongoObserver connected."));

Observable.interval(1000).take(5).subscribe(ticks);
```

### MongoObserver Internal Buffer

The `MongoObserver.connect` method prepares the capped collection used by Geyser to pass events from one process to the other. Any events passed to the MongoObserver before the database connection is ready will be saved to the observer's internal buffer and replayed once ready.

For finer-grained control (or to avoid this internal buffer), the pre-connected MongoObserver can also be obtained via subscription.

```js
import { Observable } from 'rxjs';
import { MongoObserver } from 'geyser-mongo';

MongoObserver.create({
  db: 'mongodb://localhost/my-db',
  collection: 'ticks'
}).subscribe(ticks => {
  Observable.interval(1000).take(5).subscribe(ticks);
});
```

CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.0-beta.1"></a>
# 1.0.0-beta.1 (2017-08-24)


### Features

* initial commit ([eeaa0a9](https://github.com/bitjson/geyser/commit/eeaa0a9))


src/lib.spec.ts
import { test } from "ava";
import { MongoObserver, MongoObservable } from "./";

test("library exports an observable and an observer", t => {
  t.truthy(MongoObserver);
  t.truthy(MongoObservable);
});


src/lib/mongo-observer.spec.ts
import { test } from "ava";
import { Db } from "mongodb";

import { MongoObserver } from "./mongo-observer";

const dbMock = () => (({} as any) as Db);

test("MongoObserver can be created", t => {
  const observer = MongoObserver.create({
    collection: "test-collection",
    db: dbMock()
  });
  t.truthy(observer);
});


src/lib/mongo-observable.spec.ts
import { test } from "ava";
import { Db } from "mongodb";

import { MongoObservable } from "./mongo-observable";

const dbMock = () => (({} as any) as Db);

test("MongoObservable can be created", t => {
  const observable = MongoObservable.create({
    collection: "test-collection",
    db: dbMock()
  });
  t.truthy(observable);
});


src/lib/mongo-observer.ts
import { Collection, Db, MongoClient } from "mongodb";
import {
  AsyncSubject,
  Observable,
  Observer,
  ReplaySubject,
  Subject
} from "rxjs";
import { GeyserType } from "./types";

/**
 * An observer which writes events from an observable to a MongoDB collection.
 */
export class MongoObserver implements Observer<GeyserType> {
  /**
   * Create a new MongoObserver.
   */
  public static create({
    db,
    // name of the capped collection to which this observer should write
    collection,
    // maximum number of documents in underlying capped collection
    max = 1000,
    // Maximum size of underlying capped collection
    size = 5242880,
    // number of values to buffer while waiting for MongoDB connection
    bufferSize = Number.POSITIVE_INFINITY,
    // milliseconds to buffer while waiting for MongoDB connection
    bufferTime = Number.POSITIVE_INFINITY
  }: {
    db: string | Db;
    collection: string;
    max?: number;
    size?: number;
    bufferSize?: number;
    bufferTime?: number;
  }) {
    return typeof db === "string"
      ? new MongoObserver(
          collection,
          Observable.fromPromise(MongoClient.connect(db)),
          max,
          size,
          bufferSize,
          bufferTime
        )
      : new MongoObserver(
          collection,
          Observable.of(db),
          max,
          size,
          bufferSize,
          bufferTime
        );
  }

  private connectInitiated = false;
  private isConnected = false;

  private mongoCollection: AsyncSubject<Collection> = new AsyncSubject();
  private writer: Subject<GeyserType> = new Subject();
  private buffer: ReplaySubject<GeyserType>;

  private constructor(
    private collection: string,
    private db: Observable<Db>,
    private max: number,
    private size: number,
    bufferSize: number,
    bufferTime: number
  ) {
    this.buffer = new ReplaySubject(bufferSize, bufferTime);

    // Automatically connect when the buffer begins to receive events or signals
    this.buffer.take(1).subscribe({
      complete: () => this.connect(),
      error: () => this.connect()
    });

    this.mongoCollection.subscribe(colInstance => {
      this.writeInit();
      this.writer.subscribe({
        complete: this.writeComplete,
        error: this.writeError,
        next: this.writeNext
      });
    });
  }

  /**
   * Immediately connect this MongoObserver to the database, create the 
   * collection (if it doesn't exist), and check that the collection has the 
   * expected stats.
   */
  public connect(
    // error: (error: any) => void,
    error = console.error,
    complete = () => {
      return;
    }
  ) {
    this.doSubscribe({
      complete,
      error,
      next: () => {
        return;
      }
    });
    return this;
  }

  public close() {
    // TODO: close DB connection
  }

  public next(value: GeyserType) {
    return this.isConnected ? this.writer.next(value) : this.buffer.next(value);
  }

  public complete() {
    return this.isConnected ? this.writer.complete() : this.buffer.complete();
  }

  public error(err: GeyserType) {
    return this.isConnected ? this.writer.error(err) : this.buffer.error(err);
  }

  /**
   * Subscribe to the connection of this MongoObserver to MongoDB. 
   */
  private doSubscribe(observer: Observer<MongoObserver>) {
    this.db
      .take(1)
      .switchMap(dbInstance =>
        // create the collection if it doesn't exist
        dbInstance.createCollection(this.collection, {
          capped: true,
          max: this.max,
          size: this.size
        })
      )
      .do(collectionInstance => this.mongoCollection.next(collectionInstance))
      .switchMap(collectionInstance => collectionInstance.stats())
      .map(stats => {
        let error = "";
        if (!stats.capped) {
          error = "collection must be capped.";
        }
        if (stats.maxSize !== this.size) {
          error = `expected collection maxSize of ${this
            .size}, but got ${stats.maxSize}.`;
        }
        if (stats.max !== this.max) {
          error = `expected collection max (documents) of ${this
            .max}, but got ${stats.max}.`;
        }
        if (error) {
          throw new Error(
            `MongoObserver tried to connect to an incompatible '${this
              .collection}' collection: ${error}`
          );
        }
      })
      .subscribe({
        complete: () => {
          this.isConnected = true;
          this.mongoCollection.complete();
          this.writer.subscribe(this.buffer);
          // TODO: delete the buffer now that it's been replayed
          observer.complete();
        },
        error: (err: any) => {
          observer.error(err);
        },
        next: () => {
          observer.next(this);
        }
      });
  }

  /* tslint:disable:no-console */

  private writeInit() {
    // TODO: write "stream init" document to mongo so observables can seek to latest stream
    console.log("TODO: write init");
  }

  private writeNext(val: any) {
    // TODO: write "next" document to mongo
    console.log("TODO: write next", val);
  }

  private writeComplete() {
    // TODO: write "complete" document to mongo
    console.log("TODO: write complete");
  }

  private writeError(err: any) {
    // TODO: write "error" document to mongo
    console.log("TODO: write error", err);
  }
}


src/lib/mongo-observable.ts
import { Db, MongoClient } from "mongodb";
import { Observable, Observer } from "rxjs";
import { GeyserType } from "./types";

/**
 * An observable which emits messages added to a given MongoDB collection.
 */
export class MongoObservable extends Observable<GeyserType> {
  /**
   * Create a new MongoObservable given an open MongoDB connection. 
   */
  public static create(opts: { db: Db; collection: string }) {
    return new MongoObservable(opts);
  }

  /**
   * Create a new MongoObservable by creating a new MongoDB connection. 
   */
  public static connect(opts: { uri: string; collection: string }) {
    return Observable.create((observer: Observer<any>) => {
      MongoClient.connect(opts.uri).then(db => {
        return new MongoObservable({
          collection: opts.collection,
          db
        }).subscribe(observer);
      });
    });
  }
  private constructor(opts: { db: Db; collection: string }) {
    super();
  }
}


src/examples/ticks.ts
/* tslint:disable:no-console */
import { MongoClient } from "mongodb";
import { Observable } from "rxjs";

import { MongoObservable, MongoObserver } from "../";

const url =
  process.argv[2] ||
  process.env.MONGO_URL ||
  "mongodb://localhost/geyser-testing";

MongoClient.connect(url).then(db => {
  console.log("Node.js MongoClient connected.");

  const ticksIn = MongoObserver.create({
    collection: "ticks",
    db
  });

  // const ticksOut = MongoObservable.create({
  //   collection: "ticks",
  //   db
  // });

  console.log("Testing latency through MongoDB at 1 second intervals...");

  Observable.interval(1000)
    .take(60)
    .map(i => i.toString())
    .startWith("connect")
    .do(i => console.time(i))
    .subscribe(ticksIn);

  // ticksOut.subscribe({
  //   complete: () => db.close(),
  //   next: (i: string) => console.timeEnd(i)
  // });
  // Observable.interval(10000).take(1).subscribe(() => db.close());
});
